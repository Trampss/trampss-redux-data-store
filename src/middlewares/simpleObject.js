import { SET, RESET, UPDATE } from '../actions'

export const initState = { data: undefined, initialized: false }

const init = defaultData => ({ ...initState, data: defaultData })

const reducer = (/* key */) => prefix => defaultData =>
  (state = init(defaultData), { type, payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return {
          data: payload,
          initialized: true,
        }
      case UPDATE(prefix):
        return {
          ...state,
          data: {
            ...state.data,
            ...payload,
          },
        }
      case RESET(prefix):
        return init(defaultData)
      default:
        return state
    }
  }

export default key => prefix => defaultData => (ctx = {}) => ({
  ...ctx,
  state: reducer(key)(prefix)(defaultData)(ctx.state, ctx.action),
})
