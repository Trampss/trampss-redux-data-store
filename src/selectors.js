import { at } from 'lodash'

export const getState = path => prefix => (state) => {
  let subState = state
  if (path !== undefined) subState = at(state, path)[0]

  return subState[prefix]
}

const getFactory = key => path => prefix => state => getState(path)(prefix)(state)[key]
export const getKeys = getFactory('keys')
export const getAsArray = getFactory('array')
export const getNb = getFactory('nb')
export const isInitialized = getFactory('initialized')

const getDatas = getFactory('datas')
export const get = path => prefix => state => (keys) => {
  const datas = getDatas(path)(prefix)(state)
  // All datas
  if (!keys) return datas
  // By keys
  if (Array.isArray(keys)) return keys.map(k => datas[k])
  // By key
  return datas[keys]
}

