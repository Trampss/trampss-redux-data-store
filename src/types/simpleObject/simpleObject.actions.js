export { set, SET, reset, RESET, update, UPDATE } from '../../actions'
