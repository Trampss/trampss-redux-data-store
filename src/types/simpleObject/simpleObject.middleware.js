import { SET, RESET, UPDATE } from '../../actions'

export const initState = {}
const defaultState = defaultData => defaultData !== undefined ? defaultData : initState

const reducer = (/* key */) => prefix => defaultData =>
  (state = defaultState(defaultData), { type, payload } = {}) => {
    switch (type) {
      case SET(prefix):
        return payload
      case UPDATE(prefix):
        return {
          ...state,
          ...payload,
        }
      case RESET(prefix):
        return defaultState(defaultData)
      default:
        return state
    }
  }

export default key => prefix => defaultData => (ctx = {}) => ({
  ...ctx,
  state: reducer(key)(prefix)(defaultData)(ctx.state, ctx.action),
})
