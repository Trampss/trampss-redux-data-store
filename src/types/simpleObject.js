import simpleObject from '../middlewares/simpleObject'

export default {
  middlewares: [simpleObject],
  actions: ['set', 'reset'],
  selectors: ['getState', 'isInitialized', 'get'],
}
