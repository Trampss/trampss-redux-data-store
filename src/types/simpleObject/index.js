import middleware from './simpleObject.middleware'
import * as actions from './simpleObject.actions'
import * as selectors from './simpleObject.selectors'

export default {
  middlewares: [middleware],
  actions,
  selectors,
}
