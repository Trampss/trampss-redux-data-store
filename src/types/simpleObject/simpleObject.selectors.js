import { initState } from './simpleObject.middleware'
import { getState } from '../../selectors'

const get = options => () => state => getState(options)(state)

const isInitialized = options => (state) => {
  if (options.defaultData !== undefined) return getState(options)(state) !== options.defaultData
  return get(options)()(state) !== initState
}

export { get, isInitialized }
