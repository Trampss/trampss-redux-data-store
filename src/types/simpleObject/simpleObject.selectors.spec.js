/* eslint-env jest */
import { isInitialized, get } from './simpleObject.selectors'
import { initState } from './simpleObject.middleware'

describe('selectors/simpleObject => ', () => {
  describe('isInitialized() :', () => {
    describe('without defaultData', () => {
      const withoutDefault = { name: 'o' }
      it('not initialized', () => {
        expect(isInitialized(withoutDefault)({ o: initState })).toMatchSnapshot()
      })
      it('initialized with object value', () => {
        expect(isInitialized(withoutDefault)({ o: { i: 'initialized' } })).toMatchSnapshot()
      })
      it('initialized with string value', () => {
        expect(isInitialized(withoutDefault)({ o: 'initialized' })).toMatchSnapshot()
      })
    })

    describe('defaultData should be a string', () => {
      const string = { name: 'o', defaultData: 'defaultData' }
      it('not initialized', () => {
        expect(isInitialized(string)({ o: string.defaultData })).toMatchSnapshot()
      })
      it('initialized', () => {
        expect(isInitialized(string)({ o: 'initialized' })).toMatchSnapshot()
      })
    })

    describe('defaultData should be an object', () => {
      const object = { name: 'o', defaultData: {} }
      it('not initialized', () => {
        expect(isInitialized(object)({ o: object.defaultData })).toMatchSnapshot()
      })
      it('initialized', () => {
        expect(isInitialized(object)({ o: { i: 'initialized' } })).toMatchSnapshot()
      })
    })
  })

  describe('get(): ', () => {
    describe('without defaultData', () => {
      const withoutDefault = { name: 'o' }
      it('and get the initState', () => {
        expect(get(withoutDefault)()({ o: initState })).toMatchSnapshot()
      })
    })

    describe('defaultData should be a string', () => {
      const string = { name: 'o', defaultData: '' }
      it('and get the default data', () => {
        expect(get(string)()({ o: string.defaultData })).toMatchSnapshot()
      })
      it('and get the data', () => {
        expect(get(string)()({ o: 'DATA' })).toMatchSnapshot()
      })
    })

    describe('defaultData should be an object', () => {
      const object = { name: 'o', defaultData: {} }
      it('and get the default data', () => {
        expect(get({ name: 'o' })()({ o: object.defaultData })).toMatchSnapshot()
      })
      it('and get the data', () => {
        expect(get(object)()({ o: { i: 'DATA' } })).toMatchSnapshot()
      })
    })
  })
})
