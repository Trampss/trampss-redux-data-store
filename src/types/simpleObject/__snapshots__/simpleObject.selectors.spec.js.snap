// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/simpleObject =>  get():  defaultData as a string should return the current data 1`] = `"DATA"`;

exports[`selectors/simpleObject =>  get():  defaultData as a string should return the default data 1`] = `""`;

exports[`selectors/simpleObject =>  get():  defaultData as an object should return the current data 1`] = `
Object {
  "i": "DATA",
}
`;

exports[`selectors/simpleObject =>  get():  defaultData as an object should return the default data 1`] = `Object {}`;

exports[`selectors/simpleObject =>  get():  without defaultData should return the initState 1`] = `Object {}`;

exports[`selectors/simpleObject =>  isInitialized() : defaultData should be a string should be initialized 1`] = `true`;

exports[`selectors/simpleObject =>  isInitialized() : defaultData should be a string should not be initialized 1`] = `false`;

exports[`selectors/simpleObject =>  isInitialized() : defaultData should be an object should be initialized 1`] = `true`;

exports[`selectors/simpleObject =>  isInitialized() : defaultData should be an object should not be initialized 1`] = `false`;

exports[`selectors/simpleObject =>  isInitialized() : without defaultData should be initialized -object- 1`] = `true`;

exports[`selectors/simpleObject =>  isInitialized() : without defaultData should be initialized -string- 1`] = `true`;

exports[`selectors/simpleObject =>  isInitialized() : without defaultData should not be initialized 1`] = `false`;
